codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/Synthetic.java

```diff
@@ -55,6 +54,25 @@ protected Node createNode() {
         throw new CodegenException("attempted to serialize runtime only trait");
     }
 
+    @Override
+    public boolean equals(Object other) {
+        if (other == this) {
+            return true;
+        } else if (!(other instanceof Trait)) {
+            return false;
+        } else if (hashCode() != other.hashCode()) { // take advantage of hashcode caching
+            return false;
+        } else {
+            Trait b = (Trait) other;
+            return toShapeId().equals(b.toShapeId());
+        }
+    }
+
+    @Override
+    public int hashCode() {
+        return toShapeId().hashCode() * 17 + Node.objectNode().hashCode();
```

Comment:
I would just use the hash code of the shape id, the other values are constant and thus won't add value to the final hash code.